pkg/statistics/handle/autoanalyze/refresher/refresher.go

```diff
@@ -281,10 +320,144 @@ func checkIndexesNeedAnalyze(
+	sctx sessionctx.Context,
+	tableSchema string,
+	tblInfo *model.TableInfo,
+	tblStats *statistics.Table,
+	defs []model.PartitionDefinition,
+	partitionStats map[int64]*statistics.Table,
+	autoAnalyzeRatio float64,
+	currentTs uint64,
+) *priorityqueue.TableAnalysisJob {
+	// TODO: figure out how to check the table stats version correctly for partitioned tables.
+	tableStatsVer := sctx.GetSessionVars().AnalyzeVersion
+	statistics.CheckAnalyzeVerOnTable(tblStats, &tableStatsVer)
+
+	averageChangePercentage, avgSize, minLastAnalyzeDuration, partitionNames := calculateIndicatorsForPartitions(
+		tblInfo,
+		partitionStats,
+		defs,
+		autoAnalyzeRatio,
+		currentTs,
+	)
+	partitionIndexes := checkIndexesNeedAnalyzeForPartitionedTable(
+		tblInfo,
+		defs,
+		partitionStats,
+	)
+
+	job := &priorityqueue.TableAnalysisJob{
+		TableID:              tblInfo.ID,
+		TableSchema:          tableSchema,
+		TableName:            tblInfo.Name.O,
+		TableStatsVer:        tableStatsVer,
+		ChangePercentage:     averageChangePercentage,
+		TableSize:            avgSize,
+		LastAnalysisDuration: minLastAnalyzeDuration,
+		Partitions:           partitionNames,
+		PartitionIndexes:     partitionIndexes,
+	}
+
+	return job
+}
+
+// calculateIndicatorsForPartitions calculates the average change percentage,
+// average size and average last analyze duration for the partitions that meet the threshold.
+func calculateIndicatorsForPartitions(
+	tblInfo *model.TableInfo,
+	partitionStats map[int64]*statistics.Table,
+	defs []model.PartitionDefinition,
+	autoAnalyzeRatio float64,
+	currentTs uint64,
+) (
+	avgChange float64,
+	avgSize float64,
+	avgLastAnalyzeDuration time.Duration,
+	partitionNames []string,
+) {
+	totalChangePercent := 0.0
+	totalSize := 0.0
+	count := 0.0
+	partitionNames = make([]string, 0, len(defs))
+	cols := float64(len(tblInfo.Columns))
+	totalLastAnalyzeDuration := time.Duration(0)
+
+	for _, def := range defs {
+		tblStats := partitionStats[def.ID]
+		changePercent := calculateChangePercentage(tblStats, autoAnalyzeRatio)
+		// No need to analyze the partition because it doesn't meet the threshold or stats are not loaded yet.
+		if changePercent == 0 {
+			continue
+		}
+
+		totalChangePercent += changePercent
+		// size = count * cols
+		totalSize += float64(tblStats.RealtimeCount) * cols
+		lastAnalyzeDuration := getTableLastAnalyzeDuration(tblStats, currentTs)
+		totalLastAnalyzeDuration += lastAnalyzeDuration
+		partitionNames = append(partitionNames, def.Name.O)
+		count++
+	}
+
+	avgChange = totalChangePercent / count
+	avgSize = totalSize / count
+	avgLastAnalyzeDuration = totalLastAnalyzeDuration / time.Duration(count)
+
+	return avgChange, avgSize, avgLastAnalyzeDuration, partitionNames
+}
+
+// checkIndexesNeedAnalyzeForPartitionedTable checks if the indexes of the partitioned table need to be analyzed.
+// It returns a map from index name to the names of the partitions that need to be analyzed.
```

Comment:
Please specify in the comments that this function is mainly used to check whether the **newly added index** needs to be analyzed.
Because the changePercent in this situation is zero which means that it will be left by above function.